File: app/src/main/java/de/Maxr1998/xposed/maxlock/ui/lockscreen/PinView.java
/*
 * MaxLock, an Xposed applock module for Android
 * Copyright (C) 2014-2016 Max Rumpf alias Maxr1998
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.Maxr1998.xposed.maxlock.ui.lockscreen;

import android.annotation.SuppressLint;
import android.content.Context;
import android.support.v7.view.ContextThemeWrapper;
import android.support.v7.widget.GridLayout;
import android.view.View;
import android.view.animation.AnimationUtils;
import android.widget.FrameLayout;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import de.Maxr1998.xposed.maxlock.Common;
import de.Maxr1998.xposed.maxlock.R;

import static android.view.HapticFeedbackConstants.FLAG_IGNORE_GLOBAL_SETTING;
import static android.view.HapticFeedbackConstants.FLAG_IGNORE_VIEW_SETTING;
import static android.view.HapticFeedbackConstants.LONG_PRESS;
import static android.view.HapticFeedbackConstants.VIRTUAL_KEY;

@SuppressLint("ViewConstructor")
public class PinView extends GridLayout implements View.OnClickListener, View.OnLongClickListener {

    private final LockView mLockView;
    private final List<String> values = new ArrayList<>(Arrays.asList("1", "2", "3", "4", "5", "6", "7", "8", "9", "0"));
    private final boolean hapticFeedback;
    private String okString;

    public PinView(Context context, LockView lockView) {
        super(context);
        mLockView = lockView;
        setColumnCount(3);
        setRowCount(4);
        if (mLockView.getPrefs().getBoolean(Common.SHUFFLE_PIN, false)) {
            Collections.shuffle(values);
        }
        values.add(9, null);
        if (!mLockView.getPrefs().getBoolean(Common.ENABLE_QUICK_UNLOCK, false)) {
            okString = getResources().getString(android.R.string.ok);
            if (okString.length() > 4) {
                okString = "OK";
            }
            values.add(okString);
        }

        for (int i = 0; i < values.size(); i++) {
            View v;
            if (values.get(i) != null) {
                PinDigit p = new PinDigit(new ContextThemeWrapper(getContext(), R.style.Widget_TextView_NumPadKey));
                p.setValue(values.get(i));
                p.setOnClickListener(this);
                if (values.get(i).length() == 1) {
                    p.setOnLongClickListener(this);
                }
                v = p;
            } else {
                mLockView.removeView(mLockView.findViewById(R.id.fingerprint_stub));
                v = new FrameLayout(getContext());
                v.setId(R.id.fingerprint_stub);
            }
            LayoutParams params = new LayoutParams(spec(i / 3, 1f), spec(i % 3, 1f));
            addView(v, params);
        }
        hapticFeedback = mLockView.getPrefs().getBoolean(Common.ENABLE_PATTERN_FEEDBACK, true);
    }

    @Override
    public void onClick(View v) {
        if (hapticFeedback) {
            performHapticFeedback(VIRTUAL_KEY, FLAG_IGNORE_VIEW_SETTING | FLAG_IGNORE_GLOBAL_SETTING);
        }
        String value = ((PinDigit) v).getValue();
        if (value.length() == 1) {
            mLockView.setKey(value, true);
            mLockView.appendToInput(value);
        }
        if (mLockView.getPrefs().getBoolean(Common.ENABLE_QUICK_UNLOCK, false)) {
            mLockView.checkInput();
        } else if (value.equals(okString) && !mLockView.checkInput()) {
            mLockView.setKey(null, false);
            if (hapticFeedback) {
                getHandler().postDelayed(new Runnable() {
                    @Override
                    public void run() {
                        performHapticFeedback(VIRTUAL_KEY, FLAG_IGNORE_VIEW_SETTING | FLAG_IGNORE_GLOBAL_SETTING);
                    }
                }, 120);
            }
            v.startAnimation(AnimationUtils.loadAnimation(getContext(), R.anim.shake));
        }
    }

    @Override
    public boolean onLongClick(View v) {
        mLockView.setKey(null, false);
        if (hapticFeedback)
            performHapticFeedback(LONG_PRESS, FLAG_IGNORE_VIEW_SETTING | FLAG_IGNORE_GLOBAL_SETTING);
        return true;
    }
}

File: app/src/main/java/de/Maxr1998/xposed/maxlock/ui/lockscreen/PinView.kt
/*
 * MaxLock, an Xposed applock module for Android
 * Copyright (C) 2014-2018  Max Rumpf alias Maxr1998
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package de.Maxr1998.xposed.maxlock.ui.lockscreen

import android.annotation.SuppressLint
import android.content.Context
import android.view.HapticFeedbackConstants.*
import android.view.View
import android.view.ViewManager
import android.view.animation.AnimationUtils
import android.widget.FrameLayout
import android.widget.ImageView
import androidx.appcompat.view.ContextThemeWrapper
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.gridlayout.widget.GridLayout
import de.Maxr1998.xposed.maxlock.Common
import de.Maxr1998.xposed.maxlock.R

@SuppressLint("ViewConstructor")
class PinView(ctx: Context, private val lockView: LockView) : GridLayout(ctx), View.OnClickListener, View.OnLongClickListener {
    private val values = mutableListOf("1", "2", "3", "4", "5", "6", "7", "8", "9", "0")
    private val hapticFeedback = lockView.prefs.getBoolean(Common.ENABLE_PATTERN_FEEDBACK, true)

    init {
        columnCount = 3
        rowCount = 4
        if (lockView.prefs.getBoolean(Common.SHUFFLE_PIN, false)) {
            values.shuffle()
        }
        var j = 0
        for (i in 0 until 12) {
            val v = when (i) {
                9 -> lockView.findViewById<FrameLayout>(R.id.fingerprint_stub).also {
                    (it.parent as ViewManager).removeView(it)
                }
                11 -> AppCompatImageView(ContextThemeWrapper(context, R.style.Widget_TextView_NumPadKey)).apply {
                    setImageResource(R.drawable.ic_checked_24dp)
                    scaleType = ImageView.ScaleType.CENTER
                    setOnClickListener(this@PinView)
                    setOnLongClickListener(this@PinView)
                }
                else -> AppCompatTextView(ContextThemeWrapper(context, R.style.Widget_TextView_NumPadKey)).apply {
                    text = values[j++]
                    setOnClickListener(this@PinView)
                    setOnLongClickListener(this@PinView)
                }
            }
            val params = GridLayout.LayoutParams(GridLayout.spec(i / 3, 1f), GridLayout.spec(i % 3, 1f))
            addView(v, params)
        }
    }

    override fun onClick(v: View) {
        if (hapticFeedback) {
            performHapticFeedback(VIRTUAL_KEY, FLAG_IGNORE_VIEW_SETTING or FLAG_IGNORE_GLOBAL_SETTING)
        }
        if (v is AppCompatTextView) {
            val value = v.text
            if (value.length == 1) {
                lockView.setKey(value.toString(), true)
                lockView.appendToInput(value.toString())
            }
            if (lockView.prefs.getBoolean(Common.ENABLE_QUICK_UNLOCK, false))
                lockView.checkInput()
        } else if (!lockView.checkInput()) {
            lockView.setKey(null, false)
            if (hapticFeedback) {
                handler.postDelayed({ performHapticFeedback(VIRTUAL_KEY, FLAG_IGNORE_VIEW_SETTING or FLAG_IGNORE_GLOBAL_SETTING) }, 120)
            }
            v.startAnimation(AnimationUtils.loadAnimation(context, R.anim.shake))
            lockView.handleFailedAttempt()
        }
    }

    override fun onLongClick(v: View): Boolean {
        lockView.setKey(null, false)
        if (hapticFeedback)
            performHapticFeedback(LONG_PRESS, FLAG_IGNORE_VIEW_SETTING or FLAG_IGNORE_GLOBAL_SETTING)
        return true
    }
}